howdoi/plugins/base.py

```diff
@@ -0,0 +1,166 @@
+}
+
+
+class BasePlugin():
+    def __init__(self, cache=None):
+        if cache is None:
+            cache = NullCache()
+        self.cache = cache
+
+    def _is_blocked(self, page):
+        for indicator in BLOCK_INDICATORS:
+            if page.find(indicator) != -1:
+                return True
+        return False
+
+    def _add_links_to_text(self, element):
+        hyperlinks = element.find('a')
+
+        for hyperlink in hyperlinks:
+            pquery_object = pq(hyperlink)
+            href = hyperlink.attrib['href']
+            copy = pquery_object.text()
+            if (copy == href):
+                replacement = copy
+            else:
+                replacement = "[{0}]({1})".format(copy, href)
+            pquery_object.replace_with(replacement)
+
+    def get_link_at_pos(self, links, position):
+        if not links:
+            return False
+        if len(links) >= position:
+            link = links[position - 1]
+        else:
+            link = links[-1]
+        return link
+
+    def get_text(self, element):
+        ''' return inner text in pyquery element '''
+        self._add_links_to_text(element)
+        try:
+            return element.text(squash_space=False)
+        except TypeError:
+            return element.text()
+
+    def _get_search_url(self, search_engine):
+        return SEARCH_URLS.get(search_engine, SEARCH_URLS['google'])
+
+    def _extract_links_from_bing(self, html):
+        html.remove_namespaces()
+        return [a.attrib['href'] for a in html('.b_algo')('h2')('a')]
+
+    def _extract_links_from_google(self, html):
+        return [a.attrib['href'] for a in html('.l')] or \
+            [a.attrib['href'] for a in html('.r')('a')]
+
+    def _extract_links_from_duckduckgo(self, html):
+        html.remove_namespaces()
+        links_anchors = html.find('a.result__a')
+        results = []
+        for anchor in links_anchors:
+            link = anchor.attrib['href']
+            url_obj = urlparse(link)
+            parsed_url = parse_qs(url_obj.query).get('uddg', '')
+            if parsed_url:
+                results.append(parsed_url[0])
+        return results
+
+    def _extract_links(self, html, search_engine):
+        if search_engine == 'bing':
+            return self._extract_links_from_bing(html)
+        if search_engine == 'duckduckgo':
+            return self._extract_links_from_duckduckgo(html)
+        return self._extract_links_from_google(html)
+
+    def get_proxies(self):
+        proxies = getproxies()
+        filtered_proxies = {}
+        for key, value in proxies.items():
+            if key.startswith('http'):
+                if not value.startswith('http'):
+                    filtered_proxies[key] = 'http://%s' % value
+                else:
+                    filtered_proxies[key] = value
+        return filtered_proxies
+
+    def extract(self):
+        print("Hello extract")
```

Comment:
I'm not sure why this is here.
If it's intended for it to be overriden in child classes, I'd suggest either raising a `NotImplementedError` or just leave it as it is (without the print statement).